lua_modules/mcp23017/mcp23017.lua

```diff
@@ -0,0 +1,262 @@
+local function setDefaultMode(self, bReg, iByte)
+    writeByte(self, getDirRegisterAddr(self, bReg), iByte)
+end
+
+-- reset gpio mode
+function mcp23017:reset()
+    -- all to input
+    setDefaultMode(self, self.GPA, 0xFF)
+    setDefaultMode(self, self.GPB, 0xFF)
+end
+
+-- setup device
+function mcp23017:setup(address, i2cId)
+
+    self.address = address
+    self.i2cId = i2cId
+    if (checkAddress(self) ~= true) or (checkDevice(self) ~= true) then
+        self.deviceOk = false
+        return 0
+    else
+        self.deviceOk = true
+        self:reset()
+        return 1
+    end
+end
+
+-- set mode for a pin
+function mcp23017:setMode(bReg, pin, mode)
+    if self.deviceOk == false then
+        return nil
+    end
+
+    local inReq = getDirRegisterAddr(self, bReg)
+    local inPin = checkPinIsInRange(pin)
+    local response = readByte(self, inReq)
+    local newState
+
+    if mode == self.OUTPUT then
+        newState = clearBit(response, inPin)
+    else
+        newState = setBit(response, inPin)
+    end
+
+    writeByte(self, inReq, newState)
+    return true
+end
+
+-- set pin to low or high
+function mcp23017:setPin(bReg, pin, state)
+    if self.deviceOk == false then
+        return nil
+    end
+
+    local inReq = getGPIORegisterAddr(self, bReg)
+    local inPin = checkPinIsInRange(pin)
+    local response = readByte(self, inReq)
+    local newState
+
+    if state == self.HIGH then
+        newState = setBit(response, inPin)
+    else
+        newState = clearBit(response, inPin)
+    end
+
+    writeByte(self, inReq, newState)
+    return true
+end
+
+-- read pin input
+function mcp23017:getPinState(bReg, pin)
+    if self.deviceOk == false then
+        return nil
+    end
+
+    local inReq = getGPIORegisterAddr(self, bReg)
+    local inPin = checkPinIsInRange(pin)
+    local response = readByte(self, inReq)
+    return issetBit(response, inPin)
+end
+
+local function writeToRegister(self, registerAddr, newByte)
+    if self.deviceOk == false then
+        return nil
+    end
+    return writeByte(self, registerAddr, newByte)
+end
+
+local function readFromRegister(self, registerAddr)
+    if self.deviceOk == false then
+        return nil
+    end
+    return readByte(self, registerAddr)
+end
+
+function mcp23017:writeIODIR(bReg, newByte)
+    return writeToRegister(self, getDirRegisterAddr(self, bReg), newByte)
+end
+
+function mcp23017:writeGPIO(bReg, newByte)
+    return writeToRegister(self, getGPIORegisterAddr(self, bReg), newByte)
+end
+
+function mcp23017:readGPIO(bReg)
+    return readFromRegister(self, getGPIORegisterAddr(self, bReg))
+end
+
+return function(address, i2cId)
+    local self = {}
+    setmetatable(self, mcp23017)
+
+    -- defaults
+    self.deviceOK = false
+    self.i2cId = 0
+    self.address = nil
+    self.OUTPUT = false
+    self.INPUT = true
+    self.GPA = false
+    self.GPB = true
+    self.HIGH = true
+    self.LOW = false
+
+    self:setup(address, i2cId)
```

Comment:
I'm not jazzed about this printing and always returning a table (that just happens to have `deviceOK` set to `false` on the case of an error).  I would prefer that `checkAddress` and `checkDevice` be moved to the example code and `setup` removed from the driver code.  This has the added benefit that all the `deviceOK` checks can also be dropped.